xdsl/dialects/riscv_ssa.py

```diff
@@ -0,0 +1,711 @@
+from __future__ import annotations
+
+from abc import ABC
+from dataclasses import dataclass, field
+from typing import Type, Dict, Union, Optional, Any, List, TypeVar, Annotated
+
+from xdsl.dialects.builtin import StringAttr, IntegerAttr, AnyIntegerAttr, IntAttr
+from xdsl.ir import (Operation, ParametrizedAttribute, SSAValue, Dialect,
+                     Attribute, Data, OpResult, Region)
+from xdsl.irdl import (irdl_op_definition, irdl_attr_definition, OptOpResult,
+                       VarOperand, SingleBlockRegion, OpAttr, OptOpAttr,
+                       OptOperand, Operand, ParameterDef)
+from xdsl.parser import BaseParser
+from xdsl.printer import Printer
+from xdsl.utils.exceptions import VerifyException
+
+
+@dataclass(frozen=True)
+class Register:
+    """A riscv register."""
+
+    index: int = field()
+    """The register index. Can be between 0 and 31."""
+
+    abi_names = {
+        "zero": 0,
+        "ra": 1,
+        "sp": 2,
+        "gp": 3,
+        "tp": 4,
+        "t0": 5,
+        "t1": 6,
+        "t2": 7,
+        "fp": 8,
+        "s0": 8,
+        "s1": 9,
+        "a0": 10,
+        "a1": 11,
+        "a2": 12,
+        "a3": 13,
+        "a4": 14,
+        "a5": 15,
+        "a6": 16,
+        "a7": 17,
+        "s2": 18,
+        "s3": 19,
+        "s4": 20,
+        "s5": 21,
+        "s6": 22,
+        "s7": 23,
+        "s8": 24,
+        "s9": 25,
+        "s10": 26,
+        "s11": 27,
+        "t3": 28,
+        "t4": 29,
+        "t5": 30,
+        "t6": 31
+    }
+
+    @staticmethod
+    def from_index(index: int) -> Register:
+        assert 32 > index >= 0
+        register = Register(index)
+        return register
+
+    @staticmethod
+    def from_name(name: str) -> Register:
+        if name in Register.abi_names:
+            return Register.from_index(Register.abi_names[name])
+        if name[0] == 'x' and name[1:].isnumeric():
+            return Register.from_index(int(name[1:]))
+        assert False and "register with unknown name"
+
+    def get_abi_name(self) -> str:
+        for name, index in Register.abi_names.items():
+            if index == self.index:
+                return name
+        assert False and "Register with unknown index"
+
+
+@irdl_attr_definition
+class RegisterAttr(Data[Register]):
```

Comment:
Do we need this? Could register contents not be `IntegerAttr|FloatAttr`?

Reply:
Registers are opaque things in this dialect, they don't encapsulate anything (they're just `n` bits, where `n` is platform specific)

Reply:
The advantage of it not being bits is that we can type check types but not registers. If a bunch of bits is an integer, a float, or a pointer, for example.